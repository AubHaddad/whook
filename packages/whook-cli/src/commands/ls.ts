import { extra, autoService, SPECIAL_PROPS } from 'knifecycle';
import { readdir } from 'fs';
import { readArgs } from '../libs/args.js';
import { YError } from 'yerror';
import path from 'path';
import os from 'os';
import {
  DEFAULT_IGNORED_FILES_PREFIXES,
  DEFAULT_IGNORED_FILES_SUFFIXES,
  DEFAULT_REDUCED_FILES_SUFFIXES,
  noop,
} from '@whook/whook';
import type { Service } from 'knifecycle';
import type {
  WhookCommandDefinition,
  PromptArgs,
  WhookCommandHandler,
} from '../services/promptArgs.js';
import type { LogService } from 'common-services';
import type {
  CONFIGSService,
  WhookPluginsService,
  WhookPluginsPathsService,
  ImporterService,
} from '@whook/whook';

export const definition: WhookCommandDefinition = {
  description: 'Print available commands',
  example: `whook ls`,
  arguments: {
    type: 'object',
    additionalProperties: false,
    properties: {
      verbose: {
        description: 'Output extra informations',
        type: 'boolean',
      },
    },
  },
};

export default extra(definition, autoService(initLsCommand));

async function initLsCommand({
  CONFIG,
  PROJECT_SRC,
  IGNORED_FILES_SUFFIXES = DEFAULT_IGNORED_FILES_SUFFIXES,
  IGNORED_FILES_PREFIXES = DEFAULT_IGNORED_FILES_PREFIXES,
  REDUCED_FILES_SUFFIXES = DEFAULT_REDUCED_FILES_SUFFIXES,
  WHOOK_PLUGINS,
  WHOOK_PLUGINS_PATHS,
  readDir = _readDir,
  log = noop,
  promptArgs,
  EOL = os.EOL,
  importer,
}: {
  CONFIG: CONFIGSService;
  PROJECT_SRC: string;
  IGNORED_FILES_SUFFIXES?: string[];
  IGNORED_FILES_PREFIXES?: string[];
  REDUCED_FILES_SUFFIXES?: string[];
  WHOOK_PLUGINS: WhookPluginsService;
  WHOOK_PLUGINS_PATHS: WhookPluginsPathsService;
  readDir?: typeof _readDir;
  log?: LogService;
  promptArgs: PromptArgs;
  EOL?: typeof os.EOL;
  importer: ImporterService<Service>;
}): Promise<WhookCommandHandler> {
  return async () => {
    const {
      namedArguments: { verbose },
    } = readArgs<{ verbose: boolean }>(
      definition.arguments,
      await promptArgs(),
    );
    const commandsSources = [CONFIG.name || 'project', ...WHOOK_PLUGINS];
    const commandsPaths = [PROJECT_SRC, ...WHOOK_PLUGINS_PATHS];
    const pluginsDefinitions = await Promise.all(
      commandsPaths.map(async (pluginPath, i) => {
        try {
          return {
            plugin: commandsSources[i],
            commands: (
              await Promise.all(
                [
                  ...new Set(
                    (await readDir(path.join(pluginPath, 'commands')))
                      .filter(
                        (file) =>
                          file !== '..' &&
                          file !== '.' &&
                          !IGNORED_FILES_PREFIXES.some((prefix) =>
                            file.startsWith(prefix),
                          ) &&
                          !IGNORED_FILES_SUFFIXES.some((suffix) =>
                            file.endsWith(suffix),
                          ),
                      )
                      .map((file) =>
                        REDUCED_FILES_SUFFIXES.some((suffix) =>
                          file.endsWith(suffix),
                        )
                          ? file.split('.').slice(0, -1).join('.')
                          : file,
                      ),
                  ),
                ]
                  .map((file) =>
                    path.join(pluginPath, 'commands', file + '.js'),
                  )
                  .map(async (file) => await importer(file)),
              )
            ).map(({ definition, default: initializer }) => ({
              definition,
              name: initializer[SPECIAL_PROPS.NAME].replace(/Command$/, ''),
            })),
          };
        } catch (err) {
          log(
            'debug',
            `âœ… - No commands folder found at path "${pluginPath}".`,
          );
          log('debug-stack', (err as Error).stack || 'no_stack_trace');
          return {
            plugin: commandsSources[i],
            commands: [],
          };
        }
      }),
    );

    pluginsDefinitions.forEach(({ plugin, commands }) => {
      log(
        'info',
        `${EOL}${EOL}# Provided by "${plugin}": ${
          commands.length === 0 ? 'none' : `${commands.length} commands`
        }${verbose ? EOL : ''}`,
      );
      commands.forEach(({ name, definition }) => {
        log(
          'info',
          `- ${name}: ${definition.description}${
            verbose && definition.example
              ? `${EOL}$ ${definition.example}${EOL}`
              : ''
          }`,
        );
      });
    });
  };
}

async function _readDir(dir: string): Promise<string[]> {
  return new Promise((resolve, reject) => {
    readdir(dir, (err, files) => {
      if (err) {
        reject(YError.wrap(err as Error, 'E_BAD_PLUGIN_DIR', dir));
        return;
      }
      resolve(files);
    });
  });
}
