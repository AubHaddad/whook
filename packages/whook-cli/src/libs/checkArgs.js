import Ajv from 'ajv';
import YError from 'yerror';

// TODO: Add ajv human readable error builder
export function checkArgs(schema, args) {
  const ajv = new Ajv({ coerceTypes: true });
  const validator = ajv.compile(schema);

  validator(args);

  if ((validator.errors || []).length) {
    throw new YError('E_BAD_ARGS', validator.errors);
  }
}
