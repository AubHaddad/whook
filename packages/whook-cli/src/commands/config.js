import { extra, autoService } from 'knifecycle';
import { readArgs } from '../libs/args';
import YError from 'yerror';
import miniquery from 'miniquery';

export const definition = {
  description: 'A simple program that returns the queryed config value',
  example: `whook config --name MYSQL --query 'auth.username' --default root`,
  arguments: {
    type: 'object',
    additionalProperties: false,
    required: ['name'],
    properties: {
      name: {
        description: 'Configuration name',
        type: 'string',
      },
      query: {
        description: 'Property to pickup in config (uses `miniquery`)',
        type: 'string',
      },
      default: {
        description: 'Provide a default value',
        type: 'string',
      },
    },
  },
};

export default extra(definition, autoService(initConfigCommand));

async function initConfigCommand({ CONFIGS, log, promptArgs }) {
  return async () => {
    const { name, query, default: defaultValue } = readArgs(
      definition.arguments,
      await promptArgs(),
    );

    if ('undefined' === typeof CONFIGS[name]) {
      log('error', `No config found for ${name}`);
      if ('undefined' === typeof defaultValue) {
        throw new YError('E_NO_CONFIG', name);
      }
    }

    const results = query ? miniquery(query, [CONFIGS[name]]) : [CONFIGS[name]];

    if (!results.length) {
      log('error', `Could not find any results for ${query}`);
      if ('undefined' === typeof defaultValue) {
        throw new YError('E_NO_RESULT', name, query);
      }
    }

    if (results.length > 1) {
      log(
        'error',
        `Got ${results.length} results for the "${query}" query, picking-up the first one.`,
      );
    }

    log(
      'info',
      `${JSON.stringify(results.length ? results[0] : defaultValue)}`,
    );
  };
}
