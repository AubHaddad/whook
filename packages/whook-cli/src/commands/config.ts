import { extra, autoService } from 'knifecycle';
import { readArgs } from '../libs/args';
import YError from 'yerror';
import miniquery from 'miniquery';
import { noop } from '@whook/whook';
import type { CONFIGSService } from '@whook/whook';
import type { LogService } from 'common-services';
import type {
  PromptArgs,
  WhookCommandDefinition,
  WhookCommandHandler,
} from '../services/promptArgs';

export const definition: WhookCommandDefinition = {
  description: 'A simple program that returns the queryed config value',
  example: `whook config --name MYSQL --query 'auth.username' --default root`,
  arguments: {
    type: 'object',
    additionalProperties: false,
    required: ['name'],
    properties: {
      name: {
        description: 'Configuration name',
        type: 'string',
      },
      query: {
        description: 'Property to pickup in config (uses `miniquery`)',
        type: 'string',
      },
      default: {
        description: 'Provide a default value',
        type: 'string',
      },
      pretty: {
        description: 'Pretty JSON output',
        type: 'boolean',
        default: false,
      },
    },
  },
};

export default extra(definition, autoService(initConfigCommand));

async function initConfigCommand({
  CONFIGS,
  promptArgs,
  log = noop,
}: {
  CONFIGS: CONFIGSService;
  promptArgs: PromptArgs;
  log?: LogService;
}): Promise<WhookCommandHandler> {
  return async () => {
    const {
      name,
      query,
      default: defaultValue,
      pretty,
    } = readArgs(definition.arguments, await promptArgs()) as {
      name: string;
      query: string;
      default: unknown;
      pretty: boolean;
    };

    if ('undefined' === typeof CONFIGS[name]) {
      log('error', `No config found for "${name}"`);
      if ('undefined' === typeof defaultValue) {
        throw new YError('E_NO_CONFIG', name);
      }
      log('info', `${JSON.stringify(defaultValue, null, pretty ? 2 : 0)}`);
      return;
    }

    const results = query ? miniquery(query, [CONFIGS[name]]) : [CONFIGS[name]];

    if (!results.length) {
      log('error', `Could not find any results for "${query}".`);
      if ('undefined' === typeof defaultValue) {
        throw new YError('E_NO_RESULT', name, query);
      }
    }

    if (results.length > 1) {
      log(
        'error',
        `Got ${results.length} results for the "${query}" query, picking-up the first one.`,
      );
    }

    log(
      'info',
      `${JSON.stringify(
        results.length ? results[0] : defaultValue,
        null,
        pretty ? 2 : 0,
      )}`,
    );
  };
}
