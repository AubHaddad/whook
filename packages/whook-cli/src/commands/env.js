import { extra, autoService } from 'knifecycle';
import { readArgs } from '../libs/args';
import YError from 'yerror';

const definition = {
  description: 'A command printing env values',
  example: `whook env --name NODE_ENV --default "default value"`,
  arguments: {
    type: 'object',
    additionalProperties: false,
    required: ['name'],
    properties: {
      name: {
        description: 'Environment variable name to pick-up',
        type: 'string',
      },
      default: {
        description: 'Provide a default value',
        type: 'string',
      },
    },
  },
};

export default extra(definition, autoService(initEnvCommand));

async function initEnvCommand({ ENV, log, args }) {
  return async () => {
    const { name, default: defaultValue } = readArgs(
      definition.arguments,
      args,
    );

    if (
      'undefined' === typeof ENV[name] &&
      'undefined' === typeof args.default
    ) {
      throw new YError('E_NO_ENV_VALUE', args.name);
    }

    log('info', `${ENV[name] || defaultValue}`);
  };
}
