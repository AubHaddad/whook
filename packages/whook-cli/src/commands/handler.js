import { extra, autoService } from 'knifecycle';
import YError from 'yerror';
import { checkArgs } from '../libs/checkArgs';

export const definition = {
  description: 'Runs the given server handler for testing purpose',
  example: `whook handler --name getPing --parameters '{}'`,
  arguments: {
    type: 'object',
    required: ['name'],
    properties: {
      name: {
        description: 'Handler to invoke',
        type: 'string',
      },
      parameters: {
        description: 'Parameters to invoke the handler with',
        type: 'string',
      },
    },
  },
};

export default extra(definition, autoService(initHandlerCommand));

async function initHandlerCommand({ $injector, log, args }) {
  return async () => {
    checkArgs(definition.arguments, args);

    const handlerName = args.name;
    let handlerParameters;

    try {
      handlerParameters = JSON.parse(args.parameters || '{}');
    } catch (err) {
      throw YError.wrap(err, 'E_BAD_PARAMETERS', args.parameters);
    }

    log('debug', 'handler', handlerName);
    log('debug', 'parameters', handlerParameters);

    // Maybe infer and check command definition from handler definition
    // with ajv or else

    try {
      const handler = (await $injector([handlerName]))[handlerName];
      const response = await handler(handlerParameters);

      log('info', JSON.stringify(response, null, 2));
      return response;
    } catch (err) {
      log('error', 'Got an error while running the handler.');
      log('stack', err.stack);
    }
  };
}
