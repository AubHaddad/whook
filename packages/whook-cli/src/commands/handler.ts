import { extra, autoService } from 'knifecycle';
import YError from 'yerror';
import { readArgs } from '../libs/args';
import { noop } from '@whook/whook';
import type { WhookHandler } from '@whook/whook';
import type {
  WhookCommandDefinition,
  PromptArgs,
  WhookCommandHandler,
} from '../services/promptArgs';
import type { Injector, Parameters } from 'knifecycle';
import type { LogService } from 'common-services';

export const definition: WhookCommandDefinition = {
  description: 'Runs the given server handler for testing purpose',
  example: `whook handler --name getPing --parameters '{}'`,
  arguments: {
    type: 'object',
    additionalProperties: false,
    required: ['name', 'parameters'],
    properties: {
      name: {
        description: 'Handler to invoke',
        type: 'string',
      },
      parameters: {
        description: 'Parameters to invoke the handler with',
        type: 'string',
        default: '{}',
      },
    },
  },
};

export default extra(definition, autoService(initHandlerCommand));

async function initHandlerCommand({
  $injector,
  log = noop,
  promptArgs,
}: {
  $injector: Injector<Record<string, WhookHandler>>;
  log?: LogService;
  promptArgs: PromptArgs;
}): Promise<WhookCommandHandler> {
  return async () => {
    const { name: handlerName, parameters: handlerParameters } = readArgs(
      definition.arguments,
      await promptArgs(),
    ) as { name: string; parameters: string };
    let parsedParameters: Parameters;

    try {
      parsedParameters = JSON.parse(handlerParameters || '{}');
    } catch (err) {
      throw YError.wrap(err as Error, 'E_BAD_PARAMETERS', handlerParameters);
    }

    log('debug', 'handler', handlerName);
    log('debug', 'parameters', parsedParameters);

    // Maybe infer and check command definition from handler definition
    // with ajv or else

    try {
      const handler = (await $injector([handlerName]))[handlerName];
      const response = await handler(parsedParameters, undefined);

      log('info', JSON.stringify(response, null, 2));
    } catch (err) {
      log('error', 'Got an error while running the handler.');
      log('error-stack', (err as Error).stack || 'no_stack_trace');
    }
  };
}
