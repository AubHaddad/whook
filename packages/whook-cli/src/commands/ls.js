import { extra, autoService, SPECIAL_PROPS } from 'knifecycle';
import { readdir } from 'fs';
import { readArgs } from '../libs/args';
import YError from 'yerror';
import path from 'path';
import os from 'os';

// Needed to avoid messing up babel builds ðŸ¤·
const _require = require;

export const definition = {
  description: 'Print available commands',
  example: `whook ls`,
  arguments: {
    type: 'object',
    additionalProperties: false,
    properties: {
      verbose: {
        type: 'boolean',
      },
    },
  },
};

export default extra(definition, autoService(initLsCommand));

async function initLsCommand({
  WHOOK_PLUGINS,
  WHOOK_PLUGINS_PATHS,
  readDir = _readDir,
  log,
  args,
  EOL = os.EOL,
  require = _require,
}) {
  return async () => {
    const { verbose } = readArgs(definition.arguments, args);
    const pluginsDefinitions = await Promise.all(
      WHOOK_PLUGINS_PATHS.map(async (pluginPath, i) => {
        try {
          return {
            plugin: WHOOK_PLUGINS[i],
            commands: (await readDir(path.join(pluginPath, 'commands')))
              .filter(
                file =>
                  file !== '..' && file !== '.' && !file.endsWith('.test.js'),
              )
              .map(file => path.join(pluginPath, 'commands', file))
              .map(file => require(file))
              .map(({ definition, default: initializer }) => ({
                definition,
                name: initializer[SPECIAL_PROPS.NAME].replace(/Command$/, ''),
              })),
          };
        } catch (err) {
          log('debug', 'âœ… - No commands folder found at path ${pluginPath}');
          log('stack', err.stack);
          return {
            plugin: WHOOK_PLUGINS[i],
            commands: [],
          };
        }
      }),
    );

    pluginsDefinitions.forEach(({ plugin, commands }) => {
      log(
        'info',
        `${EOL}${EOL}# Provided by "${plugin}": ${
          commands.length === 0 ? 'none' : `${commands.length} commands`
        }${verbose ? EOL : ''}`,
      );
      commands.forEach(({ name, definition }) => {
        log(
          'info',
          `- ${name}: ${definition.description}${
            verbose && definition.example
              ? `${EOL}$ ${definition.example}${EOL}`
              : ''
          }`,
        );
      });
    });
  };
}

async function _readDir(dir) {
  return new Promise((resolve, reject) => {
    readdir(dir, (err, files) => {
      if (err) {
        reject(YError.wrap('E_BAD_PLUGIN_DIR', dir));
        return;
      }
      resolve(files);
    });
  });
}
