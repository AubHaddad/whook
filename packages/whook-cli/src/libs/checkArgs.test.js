import { checkArgs } from './checkArgs';
import { definition as handlerCommandDefinition } from '../commands/handler';
import YError from 'yerror';

describe('checkArgs', () => {
  it('should work with no args', () => {
    const args = {};

    checkArgs(
      {
        type: 'object',
      },
      args,
    );

    expect({
      args,
    }).toMatchSnapshot();
  });

  it('should work with named args', () => {
    const args = {
      name: 'getPing',
      parameters: '{}',
    };

    checkArgs(handlerCommandDefinition.arguments, args);

    expect({
      args,
    }).toMatchSnapshot();
  });

  it('should report named args errors', () => {
    const args = {
      parameters: '{}',
    };

    try {
      checkArgs(handlerCommandDefinition.arguments, args);
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        args,
        errorCode: err.code,
        errorParams: err.params,
      }).toMatchSnapshot();
    }
  });
});
