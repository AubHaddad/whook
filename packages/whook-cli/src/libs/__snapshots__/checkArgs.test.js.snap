// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkArgs should report named args errors 1`] = `
Object {
  "args": Object {
    "parameters": "{}",
  },
  "errorCode": "E_BAD_ARGS",
  "errorParams": Array [
    Array [
      Object {
        "dataPath": "",
        "keyword": "required",
        "message": "should have required property 'name'",
        "params": Object {
          "missingProperty": "name",
        },
        "schemaPath": "#/required",
      },
    ],
  ],
}
`;

exports[`checkArgs should work with named args 1`] = `
Object {
  "args": Object {
    "name": "getPing",
    "parameters": "{}",
  },
}
`;

exports[`checkArgs should work with no args 1`] = `
Object {
  "args": Object {},
}
`;
